packages/@react-spectrum/tabs/src/Tabs.tsx

```diff
@@ -34,64 +36,140 @@ export function Tabs<T extends object>(props: SpectrumTabsProps<T>) {
   });
 
+  let {direction} = useLocale();
   let {styleProps} = useStyleProps(otherProps);
   let {tabListProps, tabPanelProps} = useTabs(props, state, ref);
+  let [collapse, setCollapse] = useValueEffect(false);
   let [selectedTab, setSelectedTab] = useState<HTMLElement>();
 
   useEffect(() => {
-    let tabs: HTMLElement[] = Array.from(ref.current.querySelectorAll('.' + styles['spectrum-Tabs-item']));
-    setSelectedTab(tabs.find((tab) => tab.dataset.key === state.selectedKey));
-  }, [props.children, state.selectedKey]);
+    if (ref.current) {
+      let tabs: HTMLElement[] = Array.from(ref.current.querySelectorAll('.' + styles['spectrum-Tabs-item']));
+      let selectedTab = tabs.find((tab) => tab.dataset.key === state.selectedKey);
+
+      // default to first tab if selectedTab is deleted from children
+      if (selectedTab != null) {
+        setSelectedTab(selectedTab);
+      } else if (tabs[0]) {
+        state.setSelectedKey(tabs[0].dataset.key);
+        setSelectedTab(tabs[0]);
+      }
+    }
+  }, [props.children, state, ref]);
+
+
+  let checkShouldCollapse = useCallback(() => {
+    let getFurthestPoint = (elem) => {
+      if (direction === 'rtl') {
+        return elem.getBoundingClientRect().left;
+      } else {
+        return elem.getBoundingClientRect().right;
+      }
+    };
+
+    let computeShouldCollapse = () => {
+      if (wrapperRef.current) {
+        let tabsComponent = wrapperRef.current;
+        let tabs = Array.from(ref.current.children);
+        // The last child is the TabLine so we look at the second to last child for the last Tab
+        let lastTab = tabs[tabs.length - 2];
+
+        let farEdgeTabList = getFurthestPoint(tabsComponent);
+        let farEdgeLastTab = getFurthestPoint(lastTab);
+        let shouldCollapse = direction === 'rtl' ? farEdgeLastTab < farEdgeTabList : farEdgeTabList < farEdgeLastTab;
+
+        if (shouldCollapse) {
+          return true;
+        } else {
+          return false;
+        }
+      }
+    };
+
+    if (orientation !== 'vertical') {
+      setCollapse(function* () {
+        // Make Tabs render in non-collapsed state
+        yield false;
+
+        // Compute if Tabs should collapse and update
+        yield computeShouldCollapse();
+      });
+    }
+  }, [ref, wrapperRef, direction, orientation, setCollapse]);
+
+  useEffect(() => {
+    checkShouldCollapse();
+  }, [props.children, checkShouldCollapse]);
+
+  useResizeObserver({ref: wrapperRef.current && wrapperRef, onResize: checkShouldCollapse});
+
+  let tablist = (
+    <TabList
+      {...tabListProps}
+      ref={ref}
+      orientation={orientation}
+      density={density}
+      isQuiet={isQuiet}
+      isDisabled={isDisabled}
+      state={state}
+      selectedTab={selectedTab} />
+  );
+
+  if (collapse && orientation !== 'vertical') {
+    tablist = (
+      <TabsPicker
+        {...props}
+        state={state} />
+    );
+  }
+
+  if (orientation !== 'vertical') {
```

Comment:
Mind elaborating a little bit? Is the desired final return statement something like:
```
<TabPanel>
   <CollapsibleTabList />
</TabPanel>
```
or is something like Kyle suggested in https://github.com/adobe/react-spectrum/pull/1145#discussion_r502057478 sufficient? Or I could make it so `TabList` handles adding the wrapping `spectrum-Tabs--collapsible` div (essentially moving one of these if statement from inside `Tabs` to `TabList`)

Reply:
purely psuedocode, but more like this?
```
<div>
  {orientation !== 'vertical' && collapse && <CollapsibleHorizontalTabs {...whateverProps} />}
  {!collapse && <NonCollapsingTabs ... />}
</div>
```

honestly, it's hard to write psuedocode for it because i don't really follow what's going on, haha. stuff keeps getting wrapped and replaced. I just think some small but very clear components might make this easier to read

Reply:
Ok I think I made it more understandable now